lib/rucio/tests/test_dataset_replicas.py

```diff
@@ -124,3 +127,211 @@ def test_list_dataset_replicas_archive(self):
+        assert_equal(dataset_replica['bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['length'], len(files))
+        assert_equal(dataset_replica['available_bytes'], (len(files) - 1) * file_size)
+        assert_equal(dataset_replica['available_replicas_cnt'], len(files) - 1)
+        assert_equal(str(dataset_replica['state']), 'UNAVAILABLE')
+        dataset_replica = self.db_session.query(models.CollectionReplica).filter_by(scope=self.scope, name=dataset_name, rse_id=self.rse_id2).one()
+        assert_equal(dataset_replica['bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['length'], len(files))
+        assert_equal(dataset_replica['available_bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['available_replicas_cnt'], len(files))
+        assert_equal(str(dataset_replica['state']), 'AVAILABLE')
+
+        # Set the state of the first replica on the second RSE to UNAVAILABLE -> both replicass should be unavailable
+        file_replica = self.db_session.query(models.RSEFileAssociation).filter_by(rse_id=self.rse_id2, scope=self.scope, name=files[0]['name']).one()
+        file_replica.state = constants.ReplicaState.UNAVAILABLE
+        models.UpdatedCollectionReplica(scope=self.scope, name=dataset_name, did_type=constants.DIDType.DATASET).save(session=self.db_session)
+        update_request = self.db_session.query(models.UpdatedCollectionReplica).filter(models.UpdatedCollectionReplica.scope == self.scope, models.UpdatedCollectionReplica.name == dataset_name, models.UpdatedCollectionReplica.rse_id.is_(None)).one()
+        update_collection_replica(update_request_id=update_request.id, session=self.db_session)
+        dataset_replica = self.db_session.query(models.CollectionReplica).filter_by(scope=self.scope, name=dataset_name, rse_id=self.rse_id).one()
+        assert_equal(dataset_replica['bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['length'], len(files))
+        assert_equal(dataset_replica['available_bytes'], (len(files) - 1) * file_size)
+        assert_equal(dataset_replica['available_replicas_cnt'], len(files) - 1)
+        assert_equal(str(dataset_replica['state']), 'UNAVAILABLE')
+        dataset_replica = self.db_session.query(models.CollectionReplica).filter_by(scope=self.scope, name=dataset_name, rse_id=self.rse_id2).one()
+        assert_equal(dataset_replica['bytes'], len(files) * file_size)
+        assert_equal(dataset_replica['length'], len(files))
+        assert_equal(dataset_replica['available_bytes'], (len(files) - 1) * file_size)
+        assert_equal(dataset_replica['available_replicas_cnt'], len(files) - 1)
+        assert_equal(str(dataset_replica['state']), 'UNAVAILABLE')
+
+        # Delete first replica on second RSE -> file is not longer part of dataset -> both replicas should be available
```

Comment:
Also if there are multiple replicas per file and the last file replica gets deleted, it gets also removed as a child of the dataset. If the replica update has no rse id, the columns 'bytes' and 'length' are then calculated without the replica and then the collection replica will be set as available which seems correct to me, because the file is no longer part of the dataset. But if the replica update has a rse id, the columns 'bytes' and 'length' are just copied from the last collection replica entry and it would be set as unavailable because the number of available replicas is less then the length (see line 197 where the collection replica will be unavailable).